axlearn/common/layers.py

```diff
@@ -1544,44 +1642,564 @@ def forward(self, x: Tensor, *, paddings: Tensor) -> tuple[Tensor, Tensor]:
 
-    Kernel weights have the WIO layout and in the shape of (window, 1, output_dim=input_dim).
-    Both inputs and outputs will be in the NWC layout.
+
+# Based on jax.lax.convolution._conv_transpose_padding, but ours is more intuitive.
+def conv_transpose_explicit_padding(
+    *,
+    window: Sequence[int],
+    strides: Sequence[int],
+    padding: ConvPaddingType,
+    dilation: Sequence[int],
+) -> ConvPaddingType:
+    """Convert str padding to tuple padding for conv_transpose.
+
+    Each mode follows the formulas below,
+    * SAME: (min(window-1, ceil((w+s-2)/2)), max(stride-1, floor((w+s-2)/2)))
+        pad_total = window+stride-2
+        when stride > window -> (window-1, stride-1)
+    * VALID: (window-1, max(stride-1, window-1))
+        pad_total = window+stride-2 + max(window-stride, 0)
+        when stride > window -> (window-1, stride-1)
+    * CAUSAL: (window-1, stride-1)
+        pad_total = window+stride-2
+
+    Note: output_size = input_size*stride - (window+stride-2) + pad_total
+                      = input_size*stride  <- "SAME", "CAUSAL"
+                      = input_size*stride + max(window-stride, 0)  <- "VALID"
+
+    Note: In the above equation, `window` will be replaced with `dilate_window` when dilation > 1.
+        dilate_window = (window - 1) * dilation + 1. Check conv_dilate_window()
+
+    The following illustration demonstrates how Conv Transpose operates, assuming all kernel values
+    are set to 1 for simplicity in showcasing output values.
+
+    In the window=3 and stride=1 case, this function creates outputs as follows:
+    * "SAME" padding=(1, 1)
+                    pad|       |pad
+        paddings:     0|0 0 1 1|0
+                      0 0 0  -> 0
+                        0 0 1  -> 1
+                          0 1 1  -> 2
+                            1 1 0  -> 2
+
+    * "VALID" padding=(2, 2)
+                    pad  |       |pad
+        paddings:     0 0|0 0 1 1|0 0
+                      0 0 0  -> 0
+                        0 0 0  -> 0
+                          0 0 1  -> 1
+                            0 1 1  -> 2
+                              1 1 0  -> 2
+                                1 0 0  -> 1
+
+    * "CAUSAL" padding=(2, 0)
+                    pad  |       |pad
+        paddings:     0 0|0 0 1 1|
+                      0 0 0  -> 0
+                        0 0 0  -> 0
+                          0 0 1  -> 1
+                            0 1 1  -> 2
+
+    In the window=3 and stride=2 case, this function creates outputs as follows:
+    * "SAME" padding=(2, 1)
+                    pad  |             |pad
+        paddings:     0 0|0 * 0 * 1 * 1|0
+                      0 0 0  -> 0
+                        0 0 0  -> 0
+                          0 0 0  -> 0
+                            0 0 0  -> 0
+                              0 0 1  -> 1
+                                0 1 0  -> 1
+                                  1 0 1  -> 2
+                                    0 1 0  -> 1
```

Comment:
I don't understand why stride=2 but the adjacent conv transpose windows shown are only one position apart from each other.

Reply:
Good question, in transposed convolution, stride=2 is not **window stride** but input dilation.

Given input `|0 0 1 1|`
stride=2 dilates it to `|0 * 0 * 1 * 1|`
and padding=(2,1) add padding as `0 0|0 * 0 * 1 * 1|0`.
That's all transposed convolution is about. After this, `window=3` and **window_stride=1** convolution takes place.

Reply:
Here is my understanding:

```
        inputs = jnp.asarray([0, 1, 2, 3, 4]).reshape([1, -1, 1])
        kernel = jnp.asarray([1, 2, 4]).reshape([-1, 1, 1])
        # padded_inputs = [_, _, 0, 1, 2, 3, 4]
        # windows =      [[_, _, 0],
        #                    [0, 0, 4],
        #                       [0, 2, 8]
        #                          [1, 4,12]
        #                             [2, 6,16]
        #                                [3, 8, _]
        # strided =      [[_, _, 0],
        #                       [0, 0, 4],
        #                             [0, 2, 8]
        #                                   [1, 4,12]
        #                                         [2, 6,16]
        #                                               [3, 8, _]
        # sum =                 [0, 0, 4, 2, 9, 4,14, 6,19].
        outputs = jax.lax.conv_transpose(inputs, rhs=kernel, padding=((2, 0),), strides=(2,))
        assert_allclose(outputs, jnp.asarray([0, 0, 4, 2, 9, 4, 14, 6, 19]).reshape((1, -1, 1)))
```

Note that in this case, I only see 6 convolution windows that affect the outputs (5 windows if the input length is 4). So I'm confused by the 8 windows in the above comments.